File: libs/core/tests/symbol_creation_test.cpp
/**
 * @file symbol_creation_test.cpp
 * @author Francesco Corso
 * @date 8/2/2026
 * @brief
 */
#include <catch2/catch_test_macros.hpp>

import std;
import prism.core;

using namespace prism;

TEST_CASE("Compilation merges namespace declarations", "[compilation]")
{
    auto tree = SyntaxTree::parse(R"(
        namespace A::B {
            var x: i32 = 5;
        }

        namespace A {
            namespace B {
                func f(p: i32) {}
            }
        }
    )");

    const Name assembly_name = "test";
    ImmutableArray trees{std::move(tree)};
    const auto compilation = Compilation::create(assembly_name, std::move(trees));

    auto members = compilation->assembly().global_namespace().members();
    REQUIRE(members.size() == 1);
    CHECK(members[0]->name() == Name{"A"});

    REQUIRE(members[0]->is<NamespaceSymbol>());
    members = members[0]->as<NamespaceSymbol>()->members();
    REQUIRE(members.size() == 1);

    CHECK(members[0]->name() == Name{"B"});
    members = members[0]->as<NamespaceSymbol>()->members();
    REQUIRE(members.size() == 2);

    CHECK(members[0]->name() == Name{"x"});
    CHECK(members[0]->is<VariableSymbol>());
    CHECK(members[1]->name() == Name{"f"});
    CHECK(members[1]->is<FunctionSymbol>());

    auto &x = members[0]->as<VariableSymbol>()->type();
    CHECK(x.special_type() == SpecialType::i32);

    auto &f = members[1]->as<FunctionSymbol>()->return_type();
    CHECK(f.special_type() == SpecialType::void_);

    auto &p = members[1]->as<FunctionSymbol>()->parameters()[0]->type();
    CHECK(p.special_type() == SpecialType::i32);
}

File: test/Prism.Core.Tests/SymbolCreationTest.cs
using Prism.Core.Compiling;
using Prism.Core.Symbols;
using Prism.Core.Syntax;

namespace Prism.Core.Tests;

public class SymbolCreationTest
{
    [Test]
    public void DeclarationMerging()
    {
        var tree = SyntaxTree.Parse(
            """
            namespace A::B {
                var x: i32 = 5;
            }

            namespace A {
                namespace B {
                    func f(p: i32) {}
                }
            }
            """
        );

        const string assemblyName = "test";
        var compilation = Compilation.Create(assemblyName, [tree]);

        var members = compilation.Assembly.GlobalNamespace.GetMembers();
        Assert.That(members, Has.Length.EqualTo(1));
        Assert.That(members[0].Name, Is.EqualTo("A"));

        Assert.That(members[0], Is.InstanceOf<NamespaceSymbol>());
        members = ((NamespaceSymbol)members[0]).GetMembers();
        Assert.That(members, Has.Length.EqualTo(1));
        Assert.That(members[0].Name, Is.EqualTo("B"));
        Assert.That(members[0], Is.InstanceOf<NamespaceSymbol>());

        members = ((NamespaceSymbol)members[0]).GetMembers();

        Assert.That(members, Has.Length.EqualTo(2));
        Assert.That(members[0].Name, Is.EqualTo("x"));
        Assert.That(members[0], Is.InstanceOf<VariableSymbol>());
        Assert.That(members[1].Name, Is.EqualTo("f"));
        Assert.That(members[1], Is.InstanceOf<FunctionSymbol>());

        var x = ((VariableSymbol)members[0]).Type;
        Assert.That(x.SpecialType, Is.EqualTo(SpecialType.I32));

        var f = ((FunctionSymbol)members[1]).ReturnType;
        Assert.That(f.SpecialType, Is.EqualTo(SpecialType.Void));

        var p = ((FunctionSymbol)members[1]).Parameters[0].Type;
        Assert.That(p.SpecialType, Is.EqualTo(SpecialType.I32));
    }
}
